File: 백준/1번문제/최원준/Q1717.java
// https://www.acmicpc.net/problem/
package 백준;

/*

1. 아이디어 :
    union-find를 사용합니다.

2. 시간복잡도 :
    O( n + m )
3. 자료구조 :
    배열

 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

class Q1717 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int[] par;
    static int[] rank;

    public static int find(int x) {
        if (x!= par[x]) {
            par[x] = find(par[x]);
        }
        return par[x];
    }

    public static void union(int a, int b) {
        int ra = find(a), rb = find(b);

        if (ra == rb) return;

        if (rank[ra] > rank[rb]) {
            par[rb]=ra;
        } else if (rank[ra] < rank[rb]) {
            par[ra]=rb;
        } else {
            rank[ra]++;
            par[rb] =ra;
        }
    }

    public static void main(String[] args) throws IOException {
        int[] inputs = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int N = inputs[0]+1, M = inputs[1];
        rank = new int[N];
        par = new int[N];
        for (int i=0; i<N; i++) par[i] = i;

        for (int i=0; i<M; i++) {
            inputs = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            int op = inputs[0], a = inputs[1], b = inputs[2];
            if (op == 0) { // union
                union(a, b);
            } else { // check
                System.out.println( find(a) == find(b) ? "YES" : "NO" );
            }
        }
    }
}

File: 백준/1번문제/S0okJu/Q1717.py
"""
아이디어
- Union - Find : 두 그룹을 합치거나, 특정 그룹을 찾는 등의 연산을 수행할 때 사용된다.

시간 복잡도 
- O(n)
"""
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

# 그룹의 루트 노드를 찾는다. 
def find(a):
    if a == parent[a]:
        return a
    parent[a] = find(parent[a])
    return parent[a]
# 그룹을 합친다.
# y의 루트 노드를 x 로 합치는 작업을 수행한다.
def union(a,b):
    x, y = find(a), find(b)
    if x < y:
        parent[y] = x
    else:
        parent[x] = y


n, m = map(int,input().split())
parent = [i for i in range(n+1)]

for i in range(m):
    op, a, b = map(int,input().split())
    if op == 1:
        if find(a) == find(b):
            print('YES')
        else:
            print('NO')
    else:
        union(a,b)